pkg/gadget-collection/gadgets/traceloop/gadget.go

```diff
@@ -52,138 +62,211 @@ some differences:
-	if runtime.GOARCH != "amd64" {
-		trace.Status.OperationError = "This gadget is only supported on amd64 for various reasons (https://github.com/kinvolk/traceloop/issues/28)"
+	if t.started {
+		trace.Status.State = gadgetv1alpha1.TraceStateStarted
 		return
 	}
 
-	t.cmd = exec.Command("/bin/bash", "-c", `
-		# gobpf currently uses global kprobes via debugfs/tracefs and not the Perf
-		# Event file descriptor based kprobe (Linux >=4.17). So unfortunately, kprobes
-		# can remain from previous executions. Ideally, gobpf should implement Perf
-		# Event based kprobe and fallback to debugfs/tracefs, like bcc:
-		# https://github.com/iovisor/bcc/blob/6e9b4509fc7a063302b574520bac6d49b01ca97e/src/cc/libbpf.c#L1021-L1027
-		# Meanwhile, as a workaround, delete probes manually.
-		# See: https://github.com/iovisor/gobpf/issues/223
-		echo "-:pfree_uts_ns" >> /sys/kernel/debug/tracing/kprobe_events 2>/dev/null || true
-		echo "-:pcap_capable" >> /sys/kernel/debug/tracing/kprobe_events 2>/dev/null || true
-
-		# Remove the opened files limit to avoid getting this error:
-		# error while loading "tracepoint/raw_syscalls/sys_enter" (too many open files):
-		# Indeed, traceloop creates a lof of tracers which each has maps and events:
-		# https://github.com/kinvolk/traceloop/blob/6f4efc6fca46d92c75f4ec4e6c6e1d829bdeaddf/bpf/straceback-guess-bpf.h#L27-L28
-		# So, this can generate a lof ot opened files.
-		ulimit -n hard
-
-		rm -f /run/traceloop.socket
-		exec /bin/traceloop k8s
-	`)
-	t.cmd.Stdout = os.Stdout
-	t.cmd.Stderr = os.Stderr
-	err := t.cmd.Start()
-	if err != nil {
-		trace.Status.OperationError = fmt.Sprintf("Failed to start: %s", err)
-		return
+	t.mntNsIDs = make(map[uint64]bool, 0)
+
+	genKey := func(container *containercollection.Container) string {
+		return container.Namespace + "/" + container.Podname
+	}
+
+	attachContainerFunc := func(container *containercollection.Container) error {
+		mntNsID := container.Mntns
+		key := genKey(container)
+
+		traceUnique.Lock()
+		err := traceUnique.tracer.Attach(mntNsID)
+		traceUnique.Unlock()
+		if err != nil {
+			log.Errorf("failed to attach tracer: %s", err)
+
+			return err
+		}
+
+		t.mntNsIDs[mntNsID] = true
+		log.Debugf("tracer attached for %q (%d)", key, mntNsID)
+
+		return nil
+	}
+
+	detachContainerFunc := func(container *containercollection.Container) {
+		mntNsID := container.Mntns
+		key := genKey(container)
+
+		traceUnique.Lock()
+		err := traceUnique.tracer.Detach(mntNsID)
+		traceUnique.Unlock()
+		if err != nil {
+			log.Errorf("failed to detach tracer: %s", err)
+
+			return
+		}
+
+		delete(t.mntNsIDs, mntNsID)
+		log.Debugf("tracer detached for %q (%d)", key, mntNsID)
+	}
+
+	containerEventCallback := func(event containercollection.PubSubEvent) {
+		switch event.Type {
+		case containercollection.EventTypeAddContainer:
+			attachContainerFunc(event.Container)
+		case containercollection.EventTypeRemoveContainer:
+			detachContainerFunc(event.Container)
```

Comment:
When a container is terminated, we should not delete the perf ring buffer, otherwise the user will not be able to show the trace anymore.